### PR TITLE
[GV] Add a viewer for GeckoView

### PR DESCRIPTION
**Pull request by calixteman**

### CODE REVIEW

File: gulpfile.js [hunk 1]
```diff
@@ -477,6 +478,25 @@ function createWebBundle(defines, options) {
     .pipe(replaceNonWebpackImport());
 }
 
+function createGVWebBundle(defines, options) {
+  const viewerOutputName = "viewer-geckoview.js";
+  defines = Object.assign({}, defines, { GECKOVIEW: true });
```
Comment (Snuffleupagus) [inline gulpfile.js:483]: Why not use `builder.merge` as in the rest of this file?

File: gulpfile.js [hunk 2]
```diff
@@ -1252,13 +1272,17 @@ gulp.task(
   gulp.series(
     createBuildNumber,
     function scriptingMozcentral() {
-      const defines = builder.merge(DEFINES, { MOZCENTRAL: true });
+      const defines = builder.merge(DEFINES, {
+        MOZCENTRAL: true,
+      });
```
Comment (Snuffleupagus) [inline gulpfile.js:1277]: Looks unrelated, please revert.

File: gulpfile.js [hunk 3]
```diff
@@ -1252,13 +1272,17 @@ gulp.task(
...
       return buildDefaultPreferences(defines, "mozcentral/");
     },
     function createMozcentral() {
       console.log();
       console.log("### Building mozilla-central extension");
-      const defines = builder.merge(DEFINES, { MOZCENTRAL: true });
+      const defines = builder.merge(DEFINES, {
+        MOZCENTRAL: true,
+      });
```
Comment (Snuffleupagus) [inline gulpfile.js:1285]: Looks unrelated, please revert.

File: web/viewer-geckoview.css [hunk 4]
```diff
@@ -0,0 +1,397 @@
...
+
+  --main-color: rgba(12, 12, 13, 1);
+  --body-bg-color: rgba(237, 237, 240, 1);
+  --scrollbar-color: auto;
+  --scrollbar-bg-color: auto;
+  --dialog-button-border: none;
+  --dialog-button-bg-color: rgba(12, 12, 13, 0.1);
+  --dialog-button-hover-bg-color: rgba(12, 12, 13, 0.3);
+
+  --loading-icon: url(images/loading.svg);
```
Comment (Snuffleupagus) [inline web/viewer-geckoview.css:33]: Here, and below, is this actually being used? If not, this can be removed.

File: web/viewer-geckoview.css [hunk 5]
```diff
@@ -0,0 +1,397 @@
...
+  background-color: var(--body-bg-color);
+  scrollbar-color: var(--scrollbar-color) var(--scrollbar-bg-color);
+}
+
+.hidden,
+[hidden] {
+  display: none !important;
+}
+
+#viewerContainer.pdfPresentationMode:fullscreen {
```
Comment (Snuffleupagus) [inline web/viewer-geckoview.css:90]: Is the GeckoView-viewer going to support PresentationMode at all, either now or in the future, since otherwise you probably don't need any CSS rules for that.

File: web/viewer-geckoview.html [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+    <title>PDF.js viewer</title>
+
+<!--#if MOZCENTRAL-->
+<!--#include viewer-snippet-firefox-extension.html-->
+<!--#endif-->
+
+<!--#if MOZCENTRAL-->
+<link rel="stylesheet" href="viewer.css">
+<!--#else-->
+<link rel="stylesheet" href="viewer-geckoview.css">
```
Comment (Snuffleupagus) [inline web/viewer-geckoview.html:39]: Don't you want to include this in the MOZCENTRAL build as well?

---

Is the GeckoView-viewer ever going to be used for anything except Firefox, and as a GENERIC viewer during development?
Because if not, there's probably a number of pre-processor blocks that can now be simplified or even removed; possible a bit too much copy-and-paste from the regular viewer?

File: web/app.js [hunk 7]
```diff
@@ -327,34 +327,38 @@ const PDFViewerApplication = {
...
-        case "hover":
-          viewerContainer.classList.add(`textLayer-${params.get("textlayer")}`);
-          try {
-            await loadPDFBug(this);
-            this._PDFBug.loadCSS();
-          } catch (ex) {
-            console.error(`_parseHashParameters: "${ex.message}".`);
-          }
-          break;
+    if (typeof PDFJSDev === "undefined" || PDFJSDev.test("!GECKOVIEW")) {
```
Comment (Snuffleupagus) [inline web/app.js:330]: Why do we need to do this exactly, since this code isn't even accessible in a normal Firefox browser without the `pdfjs.pdfBugEnabled` preference being manually set to `true`?
Basically, just leaving this code as-is really can't hurt as far as I'm concerned.

**Review by Snuffleupagus — COMMENTED** [review]
Just leaving a couple of comments based on a *very cursory* look.

File: web/viewer-geckoview.html (hunk 6 above)
Comment (calixteman) [reply web/viewer-geckoview.html:39]: > Don't you want to include this in the MOZCENTRAL build as well?

In m-c, we use `viewer.html` at several places, for example:
https://searchfox.org/mozilla-central/rev/17aeb39742eba71e0936ae44a51a54197100166d/toolkit/content/contentAreaUtils.js#115

so I'll map `viewer-geckoview.html` to `viewer.html` in the `jar.nm` file:
https://searchfox.org/mozilla-central/source/toolkit/components/pdfjs/jar.mn

and in order to have something consistent, I'll map the css and the js geckoview files either to keep `viewer.css` and `viewer.js`.

File: web/app.js [hunk 8]
```diff
@@ -2152,15 +2172,15 @@ function webViewerInitialized() {
 
   if (!PDFViewerApplication.supportsPrinting) {
     appConfig.toolbar.print.classList.add("hidden");
```
Comment (Snuffleupagus) [inline web/app.js:2174]: Don't you need to guard against an undefined toolbar here?
```suggestion
    appConfig.toolbar?.print.classList.add("hidden");
```

File: web/viewer-geckoview.html [hunk 9]
```diff
@@ -0,0 +1,96 @@
+<!DOCTYPE html>
```
Comment (Snuffleupagus) [inline web/viewer-geckoview.html:1]: The "regular" viewer has a BOM at the start of its HTML files, do we perhaps want to maintain that here as well?

**Review by Snuffleupagus — APPROVED** [review]
It's really good that optional chaining exists, since that really helped reduce the overall amount of changes needed in the `web/app.js` file :-)

r=me, with a couple of comments, since this looks overall reasonable to me!
